### ISSUE TITLE
Expand schemas

### ISSUE THREAD

**Issue by loganvolkers** [BigstickCarpet/swagger-parser#15]
It would be handy to provide an option to "expand" a schema to resolve things like `allOf` or `oneOf`.

It seems like there has already been some work put into this for the JSON-Editor project.

https://github.com/jdorn/json-editor/blob/92857220697d0e4265a1945b9a469962c340921e/src/core.js#L417

I envision this being similar to the "dereference" option.

If there is interest in this, I'd be happy to contribute a pull request, I just need some guidance on the best way to include this. I was basically going to either clone or extend the "deference" option.

**Comment by loganvolkers** [comment]
Original spec from here: https://github.com/jdorn/json-editor/issues/49
```
{
  "type": "object",
  "allOf": [
    {
      "properties": {
        "key1": {
          "type": "string"
        }
      },
      "required": ["key1"]
    },
    {
      "properties": {
        "key2": {
          "type": "string"
        }
      },
      "required": ["key2"]
    }
  ]
}
```
Should produce the following after it's expanded:
```
{
  "type": "object",
  "properties": {
    "key1": {
      "type": "string"
    },
    "key2": {
      "type": "string"
    }
  },
  "required": ["key1","key2"]
}
```

**Comment by BigstickCarpet** [comment]
Yeah, that seems like something we could do.  We're in the middle of a [major refactoring](https://github.com/BigstickCarpet/swagger-parser/issues/8#issuecomment-121977660) of Swagger Parser right now, and we're splitting all of the non-Swagger-specific logic into a seperate package called [JSON Schema $Ref Parser](https://github.com/BigstickCarpet/json-schema-ref-parser).  That's where this feature would go, since it deals solely with JSON Schema, not Swagger.

**Comment by loganvolkers** [comment]
This makes a lot of sense to me. I'll keep an eye on that project.

Moved this issue there: https://github.com/BigstickCarpet/json-schema-ref-parser/issues/1

**Comment by BigstickCarpet** [comment]
Cool. Thanks for moving it over.

**Comment by loganvolkers** [comment]
Update on this issue to close the loop. It looks like support for composition has been added by the swagger-api/swagger-js library.

As per their test specs `allOf` can being used for:
 - composition
 - inheritance when paired with `descriminator` (see: https://github.com/swagger-api/swagger-js/issues/469)

For my use case, we are using `allOf` for composition, and since swagger-parser doesn't support resolving `allOf` references we've started using their [Resolver](https://github.com/swagger-api/swagger-js/blob/master/lib/resolver.js) class in the meantime.

Here's an example for those also stuck trying to support composition:

```
var Resolver = require('swagger-client/lib/resolver');
new Resolver().resolve(inputSwaggerSchema, null, function (resolvedSchema, unresolved) {
   // Use the ResolvedSchema
});
```
